diff --git a/sources/app.sample/app.base/main.c b/sources/app.sample/app.base/main.c
index 1ffb732..d6987e2 100644
--- a/sources/app.sample/app.base/main.c
+++ b/sources/app.sample/app.base/main.c
@@ -107,6 +107,9 @@
    #include "ipc_sample.h"
 #endif
 
+// 241202 sy.kim
+#include "katech_can_app.h"
+
 /*
 ***************************************************************************************************
 *                                         GLOBAL VARIABLES
@@ -136,6 +139,27 @@ static void DisplayAliveLog
     void
 );
 
+// 241202 sy.kim
+static void TxCAN1_ADAS_CMD_32_50ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN1_ADAS_CMD_10_20ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN2_WHL_01_10ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN2_ADAS_PRK_20_20ms_Task
+(
+    void *                              pArg
+);
+
 /*
 ***************************************************************************************************
 *                                         FUNCTIONS
@@ -160,6 +184,23 @@ void cmain (void)
     SALRetCode_t            err;
     SALMcuVersionInfo_t     versionInfo = {0,0,0,0};
 
+    // 241202 sy.kim
+    static uint32           TxCAN1_1a0_TaskID = 0;
+    static uint32           TxCAN1_1a0_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    SALRetCode_t            err_1a0;
+
+    static uint32           TxCAN1_160_TaskID = 0;
+    static uint32           TxCAN1_160_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    SALRetCode_t            err_160;
+
+    static uint32           TxCAN2_a0_TaskID = 0;
+    static uint32           TxCAN2_a0_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    SALRetCode_t            err_a0;
+
+    static uint32           TxCAN2_165_TaskID = 0;
+    static uint32           TxCAN2_165_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    SALRetCode_t            err_165;
+
 #if ( SIC_BSP_SUPPORT_TEST_APP_CPU == 1 )
     /* To avoid affected from OS resources */
     main_dhrystone_1();
@@ -190,17 +231,53 @@ void cmain (void)
                          SAL_PRIO_APP_CFG,
                          NULL);
 
-    if (err == SAL_RET_SUCCESS)
+    // 241202 sy.kim
+    // create the TxCAN1_ADAS_CMD_32_50ms task...
+    err_1a0 = (SALRetCode_t)SAL_TaskCreate(&TxCAN1_1a0_TaskID,
+                         (const uint8 *)"TxCAN1_ADAS_CMD_32_50ms",
+                         (SALTaskFunc) &TxCAN1_ADAS_CMD_32_50ms_Task,
+                         &TxCAN1_1a0_TaskStk[0],
+                         ACFG_TASK_MEDIUM_STK_SIZE,
+                         SAL_PRIO_APP_CFG + 1,
+                         NULL);
+    // create the TxCAN1_ADAS_CMD_32_50ms task...
+    err_160 = (SALRetCode_t)SAL_TaskCreate(&TxCAN1_160_TaskID,
+                         (const uint8 *)"TxCAN1_ADAS_CMD_10_20ms",
+                         (SALTaskFunc) &TxCAN1_ADAS_CMD_10_20ms_Task,
+                         &TxCAN1_160_TaskStk[0],
+                         ACFG_TASK_MEDIUM_STK_SIZE,
+                         SAL_PRIO_APP_CFG + 2,
+                         NULL);
+    // create the TxCAN2_WHL_01_10ms task...
+    err_a0 = (SALRetCode_t)SAL_TaskCreate(&TxCAN2_a0_TaskID,
+                         (const uint8 *)"TxCAN2_WHL_01_10ms",
+                         (SALTaskFunc) &TxCAN2_WHL_01_10ms_Task,
+                         &TxCAN2_a0_TaskStk[0],
+                         ACFG_TASK_MEDIUM_STK_SIZE,
+                         SAL_PRIO_APP_CFG + 3,
+                         NULL);
+    // create the TxCAN2_WHL_01_10ms task...
+    err_165 = (SALRetCode_t)SAL_TaskCreate(&TxCAN2_165_TaskID,
+                         (const uint8 *)"TxCAN2_ADAS_PRK_20_20ms",
+                         (SALTaskFunc) &TxCAN2_ADAS_PRK_20_20ms_Task,
+                         &TxCAN2_165_TaskStk[0],
+                         ACFG_TASK_MEDIUM_STK_SIZE,
+                         SAL_PRIO_APP_CFG + 4,
+                         NULL);  
+
+    // 241202 sy.kim
+    if (err == SAL_RET_SUCCESS && err_1a0 == SAL_RET_SUCCESS && err_160 == SAL_RET_SUCCESS && err_a0 == SAL_RET_SUCCESS && err_165 == SAL_RET_SUCCESS) 
     {
         // start woring os.... never return from this function
         (void)SAL_OsStart();
     }
 }
 
-static void Lpa_Start(void)
+static void Lpa_Start(void) // Don't Active
 {
 #if ((CORTEX_M7_NP == 1) && (SIC_BSP_SUPPORT_TEST_APP_LPA_M == 1))
-	frm_lpa_init();
+	
+    frm_lpa_init();
 
 	frm_lpa_rx_setRegCB(cb_print_reg_data_info);
 	frm_lpa_rx_setDataCB(cb_print_data_frame_info);
@@ -258,13 +335,57 @@ static void Main_StartTask(void * pArg)
     AppTaskCreate();
 
     while (TRUE)
-    {  /* Task body, always written as an infinite loop.       */
+    {  /* Task body, always written as an infinite loop. */
         DisplayAliveLog();
         //mcu_printf("\n MCU Idle !!!");
         (void)SAL_TaskSleep(5000);
     }
 }
 
+
+/* 241202 sy.kim */
+static void TxCAN1_ADAS_CMD_32_50ms_Task(void * pArg)
+{
+    (void)pArg;
+
+    while (TRUE)
+    {
+        TxCAN1_ADAS_CMD_32_50ms(); // Call the function
+        (void)SAL_TaskSleep(1000);   // Sleep for 50ms
+    }
+}
+static void TxCAN1_ADAS_CMD_10_20ms_Task(void * pArg)
+{
+    (void)pArg;
+
+    while (TRUE)
+    {
+        TxCAN1_ADAS_CMD_10_20ms(); // Call the function
+        (void)SAL_TaskSleep(1000);   // Sleep for 50ms
+    }
+}
+static void TxCAN2_WHL_01_10ms_Task(void * pArg)
+{
+    (void)pArg;
+
+    while (TRUE)
+    {
+        TxCAN2_WHL_01_10ms(); // Call the function
+        (void)SAL_TaskSleep(1000);   // Sleep for 50ms
+    }
+}
+static void TxCAN2_ADAS_PRK_20_20ms_Task(void * pArg)
+{
+    (void)pArg;
+
+    while (TRUE)
+    {
+        TxCAN2_ADAS_PRK_20_20ms(); // Call the function
+        (void)SAL_TaskSleep(1000);   // Sleep for 50ms
+    }
+}
+
+
 static void AppTaskCreate(void)
 {
 #if (SIC_BSP_SUPPORT_CONSOLE == 1)
