diff --git a/sources/app.sample/app.base/main.c b/sources/app.sample/app.base/main.c
index 1ffb732..5ad0232 100644
--- a/sources/app.sample/app.base/main.c
+++ b/sources/app.sample/app.base/main.c
@@ -107,6 +107,9 @@
    #include "ipc_sample.h"
 #endif
 
+// 241202 sy.kim
+#include "katech_can_app.h"
+
 /*
 ***************************************************************************************************
 *                                         GLOBAL VARIABLES
@@ -136,6 +139,33 @@ static void DisplayAliveLog
     void
 );
 
+// 241202 sy.kim
+static void TxCAN1_ADAS_CMD_32_50ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN1_ADAS_CMD_10_20ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN1_ADAS_CMD_20_20ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN2_WHL_01_10ms_Task
+(
+    void *                              pArg
+);
+
+static void TxCAN2_ADAS_PRK_20_20ms_Task
+(
+    void *                              pArg
+);
+
+
 /*
 ***************************************************************************************************
 *                                         FUNCTIONS
@@ -160,6 +190,27 @@ void cmain (void)
     SALRetCode_t            err;
     SALMcuVersionInfo_t     versionInfo = {0,0,0,0};
 
+    // // 241202 sy.kim
+    // static uint32           TxCAN1_1a0_TaskID = 0;
+    // static uint32           TxCAN1_1a0_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    // SALRetCode_t            err_1a0;
+
+    // static uint32           TxCAN1_160_TaskID = 0;
+    // static uint32           TxCAN1_160_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    // SALRetCode_t            err_160;
+    
+    // static uint32           TxCAN1_1ea_TaskID = 0;
+    // static uint32           TxCAN1_1ea_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    // SALRetCode_t            err_1ea;
+
+    // static uint32           TxCAN2_a0_TaskID = 0;
+    // static uint32           TxCAN2_a0_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    // SALRetCode_t            err_a0;
+
+    // static uint32           TxCAN2_165_TaskID = 0;
+    // static uint32           TxCAN2_165_TaskStk[ACFG_TASK_MEDIUM_STK_SIZE];
+    // SALRetCode_t            err_165;
+
 #if ( SIC_BSP_SUPPORT_TEST_APP_CPU == 1 )
     /* To avoid affected from OS resources */
     main_dhrystone_1();
@@ -190,17 +241,65 @@ void cmain (void)
                          SAL_PRIO_APP_CFG,
                          NULL);
 
-    if (err == SAL_RET_SUCCESS)
+    // // 241202 sy.kim
+    // err_160 = (SALRetCode_t)SAL_TaskCreate(&TxCAN1_160_TaskID,
+    //                      (const uint8 *)"TxCAN1_ADAS_CMD_10_20ms",
+    //                      (SALTaskFunc) &TxCAN1_ADAS_CMD_10_20ms_Task,
+    //                      &TxCAN1_160_TaskStk[0],
+    //                      ACFG_TASK_MEDIUM_STK_SIZE,
+    //                      SAL_PRIO_APP_CFG + 1,
+    //                      NULL);
+    // err_1ea = (SALRetCode_t)SAL_TaskCreate(&TxCAN1_1ea_TaskID,
+    //                      (const uint8 *)"TxCAN1_ADAS_CMD_20_20ms",
+    //                      (SALTaskFunc) &TxCAN1_ADAS_CMD_20_20ms_Task,
+    //                      &TxCAN1_1ea_TaskStk[0],
+    //                      ACFG_TASK_MEDIUM_STK_SIZE,
+    //                      SAL_PRIO_APP_CFG + 2,
+    //                      NULL);
+    // err_a0 = (SALRetCode_t)SAL_TaskCreate(&TxCAN2_a0_TaskID,
+    //                      (const uint8 *)"TxCAN2_WHL_01_10ms",
+    //                      (SALTaskFunc) &TxCAN2_WHL_01_10ms_Task,
+    //                      &TxCAN2_a0_TaskStk[0],
+    //                      ACFG_TASK_MEDIUM_STK_SIZE,
+    //                      SAL_PRIO_APP_CFG + 3,
+    //                      NULL);
+
+
+    // err_1a0 = (SALRetCode_t)SAL_TaskCreate(&TxCAN1_1a0_TaskID,
+    //                      (const uint8 *)"TxCAN1_ADAS_CMD_32_50ms",
+    //                      (SALTaskFunc) &TxCAN1_ADAS_CMD_32_50ms_Task,
+    //                      &TxCAN1_1a0_TaskStk[0],
+    //                      ACFG_TASK_MEDIUM_STK_SIZE,
+    //                      SAL_PRIO_APP_CFG + 4,
+    //                      NULL);
+    // err_165 = (SALRetCode_t)SAL_TaskCreate(&TxCAN2_165_TaskID,
+    //                      (const uint8 *)"TxCAN2_ADAS_PRK_20_20ms",
+    //                      (SALTaskFunc) &TxCAN2_ADAS_PRK_20_20ms_Task,
+    //                      &TxCAN2_165_TaskStk[0],
+    //                      ACFG_TASK_MEDIUM_STK_SIZE,
+    //                      SAL_PRIO_APP_CFG + 5,
+    //                      NULL);  
+
+    // // 241202 sy.kim
+    // if (err == SAL_RET_SUCCESS && err_1a0 == SAL_RET_SUCCESS && err_160 == SAL_RET_SUCCESS 
+    // && err_a0 == SAL_RET_SUCCESS && err_165 == SAL_RET_SUCCESS && err_1ea == SAL_RET_SUCCESS) 
+    // {
+    //     // start woring os.... never return from this function
+    //     (void)SAL_OsStart();
+    // }
+
+    if (err == SAL_RET_SUCCESS) 
     {
         // start woring os.... never return from this function
         (void)SAL_OsStart();
     }
 }
 
-static void Lpa_Start(void)
+static void Lpa_Start(void) // Don't Active
 {
 #if ((CORTEX_M7_NP == 1) && (SIC_BSP_SUPPORT_TEST_APP_LPA_M == 1))
-	frm_lpa_init();
+	
+    frm_lpa_init();
 
 	frm_lpa_rx_setRegCB(cb_print_reg_data_info);
 	frm_lpa_rx_setDataCB(cb_print_data_frame_info);
@@ -257,14 +356,97 @@ static void Main_StartTask(void * pArg)
     /* Create application tasks */
     AppTaskCreate();
 
+    // 241205 sy.kim
+    uint32 lastTick_10ms, lastTick_20ms, lastTick_50ms;
+    uint32 currentTick;
+
+    SAL_GetTickCount(&lastTick_10ms);
+    SAL_GetTickCount(&lastTick_20ms);
+    SAL_GetTickCount(&lastTick_50ms);
+
     while (TRUE)
-    {  /* Task body, always written as an infinite loop.       */
+    {  /* Task body, always written as an infinite loop. */
         DisplayAliveLog();
+
+        // 241205 sy.kim
+        SAL_GetTickCount(&currentTick);
+        if ((currentTick - lastTick_10ms) >= 10)
+        {
+            TxCAN2_WHL_01_10ms();
+            lastTick_10ms = currentTick;
+        }
+        if ((currentTick - lastTick_20ms) >= 20)
+        {
+            TxCAN1_ADAS_CMD_10_20ms();
+            TxCAN1_ADAS_CMD_20_20ms();
+            TxCAN2_ADAS_PRK_20_20ms();
+            lastTick_20ms = currentTick;
+        }
+        if ((currentTick - lastTick_50ms) >= 50)
+        {
+            TxCAN1_ADAS_CMD_32_50ms();
+            lastTick_50ms = currentTick;
+        }
+
         //mcu_printf("\n MCU Idle !!!");
-        (void)SAL_TaskSleep(5000);
+        (void)SAL_TaskSleep(1); // 241209 sy.kim 5000 -> 1
     }
 }
 
+
+/* 241202 sy.kim */
+// static void TxCAN1_ADAS_CMD_32_50ms_Task(void * pArg) // 0x1A0
+// {
+//     (void)pArg;
+
+//     while (TRUE)
+//     {
+//         TxCAN1_ADAS_CMD_32_50ms(); // Call the function
+//         (void)SAL_TaskSleep(50);   // Sleep for 50ms
+//     }
+// }
+// static void TxCAN1_ADAS_CMD_10_20ms_Task(void * pArg) // 0x160
+// {
+//     (void)pArg;
+
+//     while (TRUE)
+//     {
+//         TxCAN1_ADAS_CMD_10_20ms(); // Call the function 
+//         (void)SAL_TaskSleep(20);   // Sleep for 20ms
+//     }
+// }
+// static void TxCAN1_ADAS_CMD_20_20ms_Task(void * pArg) // 0x1EA
+// {
+//     (void)pArg;
+
+//     while (TRUE)
+//     {
+//         TxCAN1_ADAS_CMD_20_20ms(); // Call the function
+//         (void)SAL_TaskSleep(20);   // Sleep for 20ms
+//     }
+// }
+// static void TxCAN2_WHL_01_10ms_Task(void * pArg) // 0xA0
+// {
+//     (void)pArg;
+
+//     while (TRUE)
+//     {
+//         TxCAN2_WHL_01_10ms(); // Call the function
+//         (void)SAL_TaskSleep(10);   // Sleep for 10ms
+//     }
+// }
+// static void TxCAN2_ADAS_PRK_20_20ms_Task(void * pArg) // 0x165
+// {
+//     (void)pArg;
+
+//     while (TRUE)
+//     {
+//         TxCAN2_ADAS_PRK_20_20ms(); // Call the function
+//         (void)SAL_TaskSleep(20);   // Sleep for 20ms
+//     }
+// }
+
+
 static void AppTaskCreate(void)
 {
 #if (SIC_BSP_SUPPORT_CONSOLE == 1)
