diff --git a/sources/app.sample/app.lpa/master/katech_can_app.h b/sources/app.sample/app.lpa/master/katech_can_app.h
new file mode 100644
index 0000000..88a4e56
--- /dev/null
+++ b/sources/app.sample/app.lpa/master/katech_can_app.h
@@ -0,0 +1,45 @@
+/*
+ * katech_can_app.h
+ *
+ *  Modified on: 2024. 11. 29.
+ *      Author: sy.kim
+ */
+#ifndef _KATECH_CAN_APP_H_
+#define _KATECH_CAN_APP_H_
+#include <stdint.h>
+
+#define KATECH_APP_BUFF_SIZE 2
+#define KATECH_CAN_APP_GATEWAY_MODE 0
+#define KATECH_CAN_APP_AD_MODE 1
+
+
+typedef void (*katech_rx_can_func)(uint32, uint8 *);
+
+/*
+ * tx CAN1
+ */
+// can1 - 0x1A0
+void TxCAN1_ADAS_CMD_32_50ms(void);
+// can1 - 0x1EA
+void TxCAN1_ADAS_CMD_20_20ms(void);
+// can1 - 0x160
+void TxCAN1_ADAS_CMD_10_20ms(void);
+/*
+ * tx CAN2
+ */
+// can2 - 0xA0
+void TxCAN2_WHL_01_10ms(void);
+// // can2 - 0x165
+void TxCAN2_ADAS_PRK_20_20ms(void);
+
+// void TxCAN0_DEBUG(void);
+
+// void katech_rx_can0(uint32_t can_id, uint8_t *data);
+// void katech_rx_can1(uint32_t can_id, uint8_t *data);
+// void katech_rx_can2(uint32_t can_id, uint8_t *data);
+// void katech_rx_can3(uint32_t can_id, uint8_t *data);
+// void katech_can_app_mode_set(uint8_t mode);
+
+#endif // _KATECH_CAN_APP_H_
+
+
diff --git a/sources/app.sample/app.lpa/master/katech_database.c b/sources/app.sample/app.lpa/master/katech_database.c
new file mode 100644
index 0000000..90ef369
--- /dev/null
+++ b/sources/app.sample/app.lpa/master/katech_database.c
@@ -0,0 +1,302 @@
+#include "katech_database.h"
+#include "debug.h" // 241202 sy.kim
+#include <memory.h>
+#define KATECH_APP_BUFF_SIZE 2
+
+DB_IONIQ_RxCAN0_YRS_01_10ms DB_CAN0_YRS_01_10ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_IEB_01_10ms DB_CAN0_IEB_01_10ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_WHL_01_10ms DB_CAN0_WHL_01_10ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_SAS_01_10ms DB_CAN0_SAS_01_10ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_Unknown_0x130 DB_CAN0_Unknown_0x130[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_ESC_03_20ms DB_CAN0_ESC_03_20ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_CLU_01_20ms DB_CAN0_CLU_01_20ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_CLU_02_100ms DB_CAN0_CLU_02_100ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN0_RR_C_RDR_01_50ms DB_CAN0_RR_C_RDR_01_50ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_Unknown_0x35 DB_CAN1_Unknown_0x35[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_WHL_01_10ms DB_CAN1_WHL_01_10ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_ESC_03_20ms DB_CAN1_ESC_03_20ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_CLU_01_20ms DB_CAN1_CLU_01_20ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_CLU_01_20ms2 DB_CAN1_CLU_01_20ms2[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_RR_C_RDR_02_50ms DB_CAN1_RR_C_RDR_02_50ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_SWRC_03_20ms DB_CAN1_SWRC_03_20ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN1_MFSW_01_200ms DB_CAN1_MFSW_01_200ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN2_MDPS_01_10ms DB_CAN2_MDPS_01_10ms[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN2_Unknown_0x377 DB_CAN2_Unknown_0x377[KATECH_APP_BUFF_SIZE];
+DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms DB_CAN3_ADAS_CMD_32_50ms[KATECH_APP_BUFF_SIZE];
+
+uint8_t cnt_CAN0_YRS_01_10ms;
+uint8_t cnt_CAN0_IEB_01_10ms;
+uint8_t cnt_CAN0_WHL_01_10ms;
+uint8_t cnt_CAN0_SAS_01_10ms;
+uint8_t cnt_CAN0_Unknown_0x130;
+uint8_t cnt_CAN0_ESC_03_20ms;
+uint8_t cnt_CAN0_CLU_01_20ms;
+uint8_t cnt_CAN0_CLU_02_100ms;
+uint8_t cnt_CAN0_RR_C_RDR_01_50ms;
+uint8_t cnt_CAN1_Unknown_0x35;
+uint8_t cnt_CAN1_WHL_01_10ms;
+uint8_t cnt_CAN1_ESC_03_20ms;
+uint8_t cnt_CAN1_CLU_01_20ms;
+uint8_t cnt_CAN1_CLU_01_20ms2;
+uint8_t cnt_CAN1_RR_C_RDR_02_50ms;
+uint8_t cnt_CAN1_SWRC_03_20ms;
+uint8_t cnt_CAN1_MFSW_01_200ms;
+uint8_t cnt_CAN2_MDPS_01_10ms;
+uint8_t cnt_CAN2_Unknown_0x377;
+uint8_t cnt_CAN3_ADAS_CMD_20_20ms;
+uint8_t DB_IONIQ_POS[30]={0,};
+
+
+
+
+void *DB_IONIQ_SET[20]={
+		DB_CAN0_YRS_01_10ms,
+		DB_CAN0_IEB_01_10ms,
+		DB_CAN0_WHL_01_10ms,
+		DB_CAN0_SAS_01_10ms,
+		DB_CAN0_Unknown_0x130,
+		DB_CAN0_ESC_03_20ms,
+		DB_CAN0_CLU_01_20ms,
+		DB_CAN0_CLU_02_100ms,
+		DB_CAN0_RR_C_RDR_01_50ms,
+		DB_CAN1_Unknown_0x35,
+		DB_CAN1_WHL_01_10ms,
+		DB_CAN1_ESC_03_20ms,
+		DB_CAN1_CLU_01_20ms,
+		DB_CAN1_CLU_01_20ms2,
+		DB_CAN1_RR_C_RDR_02_50ms,
+		DB_CAN1_SWRC_03_20ms,
+		DB_CAN1_MFSW_01_200ms,
+		DB_CAN2_MDPS_01_10ms,
+		DB_CAN2_Unknown_0x377,
+		DB_CAN3_ADAS_CMD_32_50ms
+};
+
+
+void *katech_data_base_get_addr(uint32_t name_db, uint8_t pos)
+{
+	switch(name_db)
+	{
+		case KATECH_NAME_DB_IONIQ_RxCAN0_YRS_01_10ms:
+			return &DB_CAN0_YRS_01_10ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_IEB_01_10ms:
+			return &DB_CAN0_IEB_01_10ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_WHL_01_10ms:
+			return &DB_CAN0_WHL_01_10ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_SAS_01_10ms:
+			return &DB_CAN0_SAS_01_10ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_Unknown_0x130:
+			return &DB_CAN0_Unknown_0x130[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_ESC_03_20ms:
+			return &DB_CAN0_ESC_03_20ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_CLU_01_20ms:
+			return &DB_CAN0_CLU_01_20ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_CLU_02_100ms:
+			return &DB_CAN0_CLU_02_100ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN0_RR_C_RDR_01_50ms:
+			return &DB_CAN0_RR_C_RDR_01_50ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_Unknown_0x35:
+			return &DB_CAN1_Unknown_0x35[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_WHL_01_10ms:
+			return &DB_CAN1_WHL_01_10ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_ESC_03_20ms:
+			return &DB_CAN1_ESC_03_20ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_CLU_01_20ms:
+			return &DB_CAN1_CLU_01_20ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_CLU_01_20ms2:
+			return &DB_CAN1_CLU_01_20ms2[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_RR_C_RDR_02_50ms:
+			return &DB_CAN1_RR_C_RDR_02_50ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_SWRC_03_20ms:
+			return &DB_CAN1_SWRC_03_20ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN1_MFSW_01_200ms:
+			return &DB_CAN1_MFSW_01_200ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN2_MDPS_01_10ms:
+			return &DB_CAN2_MDPS_01_10ms[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN2_Unknown_0x377:
+			return &DB_CAN2_Unknown_0x377[pos];
+			break;
+		case KATECH_NAME_DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms:
+			return &DB_CAN3_ADAS_CMD_32_50ms[pos];
+			break;
+
+		default:
+			break;
+	}
+
+
+}
+
+void katech_database_update(uint32_t name_db)
+{
+	mcu_printf("241202 sy.kim katech_database_update test\n");
+	uint8_t pos;
+	pos=(DB_IONIQ_POS[name_db]+1)%KATECH_APP_BUFF_SIZE;
+	DB_IONIQ_POS[name_db]=pos;
+}
+
+void *katech_database_write_addr_get(uint32_t name_db)
+{	
+	mcu_printf("241202 sy.kim katech_database_write_addr_get test\n");
+	uint8_t pos;
+	pos=(DB_IONIQ_POS[name_db]+1)%KATECH_APP_BUFF_SIZE;
+	return katech_data_base_get_addr(name_db,pos);
+}
+void *katech_database_read_addr_get(uint32_t name_db)
+{
+	uint8_t pos;
+	pos=(DB_IONIQ_POS[name_db]);
+	return katech_data_base_get_addr(name_db,pos);
+
+}
+float data_baseAX;
+void to_autoware(KATECH_CONTROL_V2A *data)
+{
+	DB_IONIQ_RxCAN0_Unknown_0x130 *dbaddr;
+	DB_IONIQ_RxCAN0_YRS_01_10ms *this_DB_CAN0_YRS_01_10ms;
+	DB_IONIQ_RxCAN0_WHL_01_10ms *whl_db;
+	DB_IONIQ_RxCAN0_SAS_01_10ms *sas_db;
+	DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms *adas_db;
+	DB_IONIQ_RxCAN0_CLU_01_20ms *clu_db;
+	DB_IONIQ_RxCAN1_Unknown_0x35 *unknown_db;
+	DB_IONIQ_RxCAN1_MFSW_01_200ms *mfsw_db;
+	DB_IONIQ_RxCAN2_MDPS_01_10ms *mdps_db;
+
+
+	dbaddr = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_Unknown_0x130);
+	data->GearReport = dbaddr->GearPosition;
+
+
+	this_DB_CAN0_YRS_01_10ms = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_YRS_01_10ms);
+	data->HeadingRate = this_DB_CAN0_YRS_01_10ms->YawRate;
+
+	clu_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_CLU_01_20ms);
+	data->LateralVelocity = 0;
+	data->LongitudinalVelocity = clu_db->DisplaySpd;
+
+	sas_db=katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_SAS_01_10ms);
+	data->SteeringTireAngle=sas_db->StrAngle;
+
+
+
+	mfsw_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN1_MFSW_01_200ms);
+	if(mfsw_db->LeftTurnFlag==1)
+		data->TurnIndicatorsReport=1;
+	else if(mfsw_db->RightTurnFlag==1)
+		data->TurnIndicatorsReport=2;
+	else
+		data->TurnIndicatorsReport=0;
+
+	mdps_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN2_MDPS_01_10ms);
+	data->ControlMode=mdps_db->MDPS_PaModeSta;
+
+
+	data->energy_level=100;
+	data->HazardLightsReport=1;
+
+}
+
+
+void katech_get_vehicle_interface_info(KATECH_VEHICLE_INTERFACE_INFO *data)
+{
+	DB_IONIQ_RxCAN0_Unknown_0x130 *dbaddr;
+	DB_IONIQ_RxCAN0_YRS_01_10ms *this_DB_CAN0_YRS_01_10ms;
+	DB_IONIQ_RxCAN0_WHL_01_10ms *whl_db;
+	DB_IONIQ_RxCAN0_SAS_01_10ms *sas_db;
+	DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms *adas_db;
+	DB_IONIQ_RxCAN0_CLU_01_20ms *clu_db;
+	DB_IONIQ_RxCAN1_Unknown_0x35 *unknown_db;
+	DB_IONIQ_RxCAN1_MFSW_01_200ms *mfsw_db;
+	DB_IONIQ_RxCAN2_MDPS_01_10ms *mdps_db;
+
+
+	dbaddr = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_Unknown_0x130);
+	data->GearState = dbaddr->GearPosition;
+
+
+	this_DB_CAN0_YRS_01_10ms = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_YRS_01_10ms);
+	data->YawRate = this_DB_CAN0_YRS_01_10ms->YawRate;
+	data->LateralAccel = this_DB_CAN0_YRS_01_10ms->LatAccel;
+	data->LongitudinalAccel = this_DB_CAN0_YRS_01_10ms->LongAccel;
+
+	whl_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_WHL_01_10ms);
+	data->WheelSpeedFL = whl_db->WheelSpdFL;
+	data->WheelSpeedFR = whl_db->WheelSpdFR;
+	data->WheelSpeedRL = whl_db->WheelSpdRL;
+	data->WheelSpeedRR = whl_db->WheelSpdRR;
+	data->WheelPulseFL = whl_db->WheelPulseFL;
+	data->WheelPulseFR = whl_db->WheelPulseFR;
+	data->WheelPulseRL = whl_db->WheelPulseRL;
+	data->WheelPulseRR = whl_db->WheelPulseRR;
+
+	sas_db=katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_SAS_01_10ms);
+	data->HandleAngle=sas_db->StrAngle;
+	data->HandleSpd=sas_db->StrSpeed;
+
+	adas_db=katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms);
+	data->Distance = adas_db->FrontSCC_ObjDstVal;
+	data->Distance = data->Distance*0.1;
+	data->RelativeVelocity = adas_db->RadarSCC_ObjRelSpdVal;
+
+	clu_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN0_CLU_01_20ms);
+	data->ClusterVelocity = clu_db->DisplaySpd;
+
+	unknown_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN1_Unknown_0x35);
+	data->AccelState = unknown_db->AccelAct;
+
+	mfsw_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN1_MFSW_01_200ms);
+	data->LeftTurnSwitch = mfsw_db->LeftTurnFlag;
+	data->RightTurnSwitch = mfsw_db->RightTurnFlag;
+
+	mdps_db = katech_database_read_addr_get(KATECH_NAME_DB_IONIQ_RxCAN2_MDPS_01_10ms);
+	data->MDPSmode=mdps_db->MDPS_PaModeSta;
+}
+
+KATECH_CONTROL_A2V katech_vehicle_interface_a2v[KATECH_APP_BUFF_SIZE];
+uint8_t cnt_katech_vehicle_interface_control=0;
+void katech_set_vehicle_interface_control(KATECH_CONTROL_A2V *data)
+{
+	int pos;
+
+	pos= (cnt_katech_vehicle_interface_control+1)%KATECH_APP_BUFF_SIZE;
+	memcpy(&katech_vehicle_interface_a2v[pos],data,sizeof(KATECH_CONTROL_A2V));
+	cnt_katech_vehicle_interface_control = pos;
+}
+
+void katech_get_vehicle_interface_control(KATECH_CONTROL_A2V *data)
+{
+	int pos;
+
+	pos= (cnt_katech_vehicle_interface_control);
+	data->Accel=katech_vehicle_interface_a2v[pos].Accel;
+	data->SteeringAngle=katech_vehicle_interface_a2v[pos].SteeringAngle;
+}
+
+KATECH_CONTROL_A2V *katech_get_addr_vehicle_interface_control(void)
+{
+	int pos;
+
+	pos= (cnt_katech_vehicle_interface_control);
+	return &katech_vehicle_interface_a2v[pos];
+
+}
+
+
diff --git a/sources/app.sample/app.lpa/master/katech_database.h b/sources/app.sample/app.lpa/master/katech_database.h
new file mode 100644
index 0000000..3f2e1e7
--- /dev/null
+++ b/sources/app.sample/app.lpa/master/katech_database.h
@@ -0,0 +1,279 @@
+/*
+ * katech_database.h
+ *
+ *  Modified on: 2024. 11. 29.
+ *      Author: sy.kim
+ */
+
+#ifndef KATECH_DATABASE_H_
+#define KATECH_DATABASE_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+#define KATECH_NAME_DB_IONIQ_RxCAN0_YRS_01_10ms 0
+typedef struct _KATECH_DB_IONIQ_RxCAN0_YRS_01_10ms_{
+	float LatAccel;
+	float LongAccel;
+	float YawRate;
+}DB_IONIQ_RxCAN0_YRS_01_10ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_IEB_01_10ms 1
+typedef struct _KATECH_DB_IONIQ_RxCAN0_IEB_01_10ms_{
+	uint8_t BrakeAct;
+	float BrakePedalValue;
+}DB_IONIQ_RxCAN0_IEB_01_10ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_WHL_01_10ms 2
+typedef struct _KATECH_DB_IONIQ_RxCAN0_WHL_01_10ms_{
+	float WheelSpdFL;
+	float WheelSpdFR;
+	float WheelSpdRL;
+	float WheelSpdRR;
+	float WheelPulseFL;
+	float WheelPulseFR;
+	float WheelPulseRL;
+	float WheelPulseRR;
+
+	float AvgWheelSpeed;
+}DB_IONIQ_RxCAN0_WHL_01_10ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_SAS_01_10ms 4
+typedef struct _KATECH_DB_IONIQ_RxCAN0_SAS_01_10ms_{
+	float StrAngle;
+	float StrSpeed;
+}DB_IONIQ_RxCAN0_SAS_01_10ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_Unknown_0x130 5
+typedef struct _KATECH_DB_IONIQ_RxCAN0_Unknown_0x130_{
+	uint8_t GearPosition;
+	uint8_t ADCMGearPosition;
+}DB_IONIQ_RxCAN0_Unknown_0x130;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_ESC_03_20ms 6
+typedef struct _KATECH_DB_IONIQ_RxCAN0_ESC_03_20ms_{
+	uint8_t ParkingBrk;
+}DB_IONIQ_RxCAN0_ESC_03_20ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_CLU_01_20ms 7
+typedef struct _KATECH_DB_IONIQ_RxCAN0_CLU_01_20ms_{
+	uint8_t DisplaySpd;
+}DB_IONIQ_RxCAN0_CLU_01_20ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_CLU_02_100ms 8
+typedef struct _KATECH_DB_IONIQ_RxCAN0_CLU_02_100ms_{
+	uint8_t FuelLevel;
+	float AvgFuelConsum;
+	float Odometer;
+	uint16_t DTE;
+}DB_IONIQ_RxCAN0_CLU_02_100ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN0_RR_C_RDR_01_50ms 9
+typedef struct _KATECH_DB_IONIQ_RxCAN0_RR_C_RDR_01_50ms_{
+	uint8_t RadarFusion;
+}DB_IONIQ_RxCAN0_RR_C_RDR_01_50ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_Unknown_0x35 10
+typedef struct _KATECH_DB_IONIQ_RxCAN1_Unknown_0x35_{
+	uint8_t AccelAct;
+}DB_IONIQ_RxCAN1_Unknown_0x35;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_WHL_01_10ms 11
+typedef struct _KATECH_DB_IONIQ_RxCAN1_WHL_01_10ms_{
+	float WHL_SpdFLVal;
+	float WHL_SpdFRVal;
+	float WHL_SpdRLVal;
+	float WHL_SpdRRVal;
+	float Avg_WhlSpd;
+
+}DB_IONIQ_RxCAN1_WHL_01_10ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_ESC_03_20ms 12
+typedef struct _KATECH_DB_IONIQ_RxCAN1_ESC_03_20ms_{
+	uint8_t PrkBrakeAct;
+
+}DB_IONIQ_RxCAN1_ESC_03_20ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_CLU_01_20ms 13
+typedef struct _KATECH_DB_IONIQ_RxCAN1_CLU_01_20ms_{
+	uint8_t CurrentSpeed;
+
+}DB_IONIQ_RxCAN1_CLU_01_20ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_CLU_01_20ms2 14
+typedef struct _KATECH_DB_IONIQ_RxCAN1_CLU_01_20ms2_{
+	uint8_t LaneQualityL;
+	uint8_t LaneQualityR;
+	float LanePositionL;
+	float LanePositionR;
+	float LaneWidthEstimation;
+	float CurvatureL;
+	float CurvatureRateL;
+	float CurvatureR;
+	float CurvatureRateR;
+	float HeadingAngleL;
+	float HeadingAngleR;
+	uint8_t CAMStatus;
+	int16_t ThisLanePositionL;
+	int16_t ThisLanePositionR;
+	int16_t ThisHeadingAngleL;
+	int16_t ThisHeadingAngleR;
+	int16_t ThisCurvatureL;
+	int16_t ThisCurvatureR;
+	int16_t ThisCurvatureRateL;
+	int16_t ThisCurvatureRateR;
+}DB_IONIQ_RxCAN1_CLU_01_20ms2;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_RR_C_RDR_02_50ms 15
+typedef struct _KATECH_DB_IONIQ_RxCAN1_RR_C_RDR_02_50ms_{
+	uint8_t LKALHLnWaringState;
+	uint8_t LKARHLnWaringState;
+}DB_IONIQ_RxCAN1_RR_C_RDR_02_50ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_SWRC_03_20ms 16
+typedef struct _KATECH_DB_IONIQ_RxCAN1_SWRC_03_20ms_{
+	uint8_t SWRC_CrsMainSwSta;
+	uint8_t SWRC_PlusMinusSwSta;
+	uint8_t PlusSW;
+	uint8_t MinusSW;
+	uint8_t PauseResumeSW;
+	uint8_t SCC_OpSta;
+	uint8_t SCC_MainOnOffSta;    // 0 : OFF / 1 : ON
+	char SCC_StayTune;
+	uint8_t SCC_TrgtSpdSetVal;
+	uint16_t SCC_AccelReqVal;
+	uint16_t SCC_AccelReqRawVal; // offset = -10.23 / factor = 0.01
+	uint8_t CurrentSpeed;
+
+}DB_IONIQ_RxCAN1_SWRC_03_20ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN1_MFSW_01_200ms 17
+typedef struct _KATECH_DB_IONIQ_RxCAN1_MFSW_01_200ms_{
+	uint8_t TurnSigLeft;
+	uint8_t TurnSigRight;
+	uint8_t LeftTurnFlag;
+	uint8_t RightTurnFlag;
+}DB_IONIQ_RxCAN1_MFSW_01_200ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN2_MDPS_01_10ms 18
+typedef struct _KATECH_DB_IONIQ_RxCAN2_MDPS_01_10ms_{
+	uint8_t MDPS_PaModeSta;
+}DB_IONIQ_RxCAN2_MDPS_01_10ms;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN2_Unknown_0x377 19
+typedef struct _KATECH_DB_IONIQ_RxCAN2_Unknown_0x377_{
+	uint8_t SPASStatus;
+}DB_IONIQ_RxCAN2_Unknown_0x377;
+
+#define KATECH_NAME_DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms 20
+typedef struct _KATECH_DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms_{
+	uint8_t SCC_ObjSta;
+	uint8_t SCC_VehStpReq;
+	uint8_t RADARStatus;
+	uint16_t FrontSCC_ObjDstVal;
+	float RadarSCC_ObjRelSpdVal;
+}DB_IONIQ_RxCAN3_ADAS_CMD_32_50ms;
+
+#pragma pack(push,4)
+typedef struct {
+	uint32_t ts_sec;
+	uint32_t ts_nsec;
+}KATEHC_VEHICLE_INTERFACE_MSG_INFO;
+#pragma pack(pop)
+
+#pragma pack(push,4)
+typedef struct {
+	uint16_t count;
+	uint8_t CurrMode;
+	float GearState;
+	float YawRate;
+	float LateralAccel;
+	float LongitudinalAccel;
+	float WheelSpeedFL;
+	float WheelSpeedFR;
+	float WheelSpeedRL;
+	float WheelSpeedRR;
+	float WheelPulseFL;
+	float WheelPulseFR;
+	float WheelPulseRL;
+	float WheelPulseRR;
+	float HandleAngle;
+	float HandleSpd;
+	float Distance;
+	float RelativeVelocity;
+	float ClusterVelocity;
+	float AccelState;
+	float LeftTurnSwitch;
+	float RightTurnSwitch;
+	unsigned char MDPSmode;
+}KATECH_VEHICLE_INTERFACE_INFO;
+#pragma pack(pop)
+
+#pragma pack(push,4)
+typedef struct{
+	uint16_t count;
+	uint8_t TargetMode;
+	//uint8_t TargetSpeed;
+	float Handle; // 4
+	float Ax; //SCC_AccelReqVal m/s^2
+
+	float TargetSteeringAngle;
+	float TargetSteeringAngleSpeed;
+	float TargetSpeed;
+	float Accel;
+	float Jerk;
+
+	uint8_t GearCommand;
+	uint8_t GateMode;
+	bool emergency;
+	uint8_t TurnIndicatorsCommand;
+	uint8_t HazardLightsCommand;
+}KATECH_VEHICLE_INTERFACE_CONTROL;
+#pragma pack(pop)
+
+
+#pragma pack(push,4)
+typedef struct{
+	uint16_t count;
+	uint8_t TargetMode;
+	//uint8_t TargetSpeed;
+
+	float SteeringAngle;
+	float SteeringAngleSpeed;
+	float Speed;
+	float Accel;
+	float Jerk;
+
+	uint8_t GearCommand;
+	uint8_t GateMode;
+	bool emergency;
+	uint8_t TurnIndicatorsCommand;
+	uint8_t HazardLightsCommand;
+}KATECH_CONTROL_A2V;
+#pragma pack(pop)
+
+#pragma pack(push,4)
+typedef struct{
+	uint16_t count;
+	//uint8_t TargetSpeed;
+	uint8_t ControlMode;
+	float energy_level;
+	uint8_t GearReport;
+	uint8_t HazardLightsReport;
+	uint8_t TurnIndicatorsReport;
+	float SteeringTireAngle;
+	float LongitudinalVelocity;
+	float LateralVelocity;
+	float HeadingRate;
+}KATECH_CONTROL_V2A;
+#pragma pack(pop)
+
+//#pragma pack(pop)
+void katech_database_update(uint32_t name_db);
+void *katech_database_read_addr_get(uint32_t name_db);
+void *katech_database_write_addr_get(uint32_t name_db);
+void katech_get_vehicle_interface_info(KATECH_VEHICLE_INTERFACE_INFO *data);
+KATECH_CONTROL_A2V *katech_get_addr_vehicle_interface_control(void);
+void katech_set_vehicle_interface_control(KATECH_CONTROL_A2V *data);
+void katech_get_vehicle_interface_control(KATECH_CONTROL_A2V *data);
+void to_autoware(KATECH_CONTROL_V2A *data);
+#endif /* KATECH_DATABASE_H_ */
