--- axon-ipc-dev.c	2024-10-10 15:52:39.326724892 +0900
+++ axon-ipc-dev-v2.c	2024-11-06 20:00:44.851148387 +0900
@@ -31,6 +31,7 @@
 #include <stdint.h>
 #include <stdlib.h>
 #include <poll.h>
+#include <sys/time.h>
 
 #include "axon_ipc.h"
 #include "axon-ipc-ctrl-log.h"
@@ -51,7 +52,14 @@
 char writeBuf[512] = {0,};
 
 uint8_t CetracSend[512];
-
+// 241007 sy.kim
+uint8_t RX_can_port;
+uint32_t RX_can_ID;
+uint8_t RX_data_len;
+uint8_t RX_Data[64U];
+// 241104 sy.kim
+struct timeval start, end;
+long elapsed_time;
 
 void parse_data_frame(const uint8_t *buffer, uint16_t data_length)
 {
@@ -113,7 +121,7 @@
 		data_frame.ts_us_high = rx_timeStamp_us_H;
 		data_frame.ts_us_low = rx_timeStamp_us_L;
 		data_frame.ts_ns = rx_timeStamp_ns;
-
+		
 		for(i = 15U; i < data_length; i++)
 		{
 			data_frame.Data[i-15U] = buffer[i];
@@ -162,6 +170,11 @@
 			PORT_NAME[data_frame.port],
 			data_frame.data_len);
 
+		// 241007 sy.kim
+		RX_can_ID = data_frame.ID;
+		RX_can_port = data_frame.port - 5;
+		RX_data_len = data_frame.data_len;
+
 		switch(data_frame.ftype)
 		{
 			case FRAME_TYPE_CAN_BASE:
@@ -188,6 +201,7 @@
 		for(i = 0U; i < data_frame.data_len; i++)
 		{
 			printf("%02x ", data_frame.Data[i]);
+			RX_Data[i] = data_frame.Data[i]; // 241007 sy.kim
 		}
 		printf("\n");
 	}
@@ -222,7 +236,7 @@
         return -1;
     }
 
-#if 1
+#if 1	// A65E RX
 	fds[0].fd = fd0;
 	fds[0].events = POLLIN;
 
@@ -387,14 +401,16 @@
 int32_t ipc_open3(void)
 {
 	int32_t ret = 0;
-	uint32_t mode = IPC_MODE_0_MBOX;
+	// 241104 sy.kim
+	// uint32_t mode = IPC_MODE_0_MBOX; 
+	uint32_t mode = IPC_MODE_0_SHM;
 	char readBuf[512] = {0};
 	struct pollfd fds[1];
 	int timeout = -1;
 	ssize_t read_size;
 	int ipc_read_packet_size = 512;
-
-	fd3 = open(AXON_IPC_CMN_FILE, O_RDWR); // Driver open per CM7 Core
+	// 241008 sy.kim
+	fd3 = open(AXON_IPC_CMN_FILE, O_RDWR | O_NONBLOCK); // Driver open per CM7 Core
 
 	if(fd3 < 0)
     {
@@ -416,7 +432,9 @@
 int32_t ipc_oprd3(void)
 {
 	int32_t ret = 0;
-	uint32_t mode = IPC_MODE_0_MBOX;
+	// 241104 sy.kim
+	// uint32_t mode = IPC_MODE_0_MBOX;
+	uint32_t mode = IPC_MODE_0_SHM;
 	char readBuf[512] = {0};
 	struct pollfd fds[1];
 	int timeout = -1;
@@ -427,8 +445,8 @@
 	uint16_t cmd2;
 	uint16_t IPC_datalen;
 	uint8_t IPC_data[132];
-
-	fd3 = open(AXON_IPC_CMN_FILE, O_RDWR);
+	// 241008 sy.kim
+	fd3 = open(AXON_IPC_CMN_FILE, O_RDWR | O_NONBLOCK);
 
 	if(fd3 < 0)
     {
@@ -450,7 +468,7 @@
 
 	while (1) {
 		printf("Waiting for data...\n");
-        int poll_ret = poll(fds, 1, timeout); // Linux poll 함수 사용하여 IPC data 수신될 때까지 지속대기
+        int poll_ret = poll(fds, 1, timeout); // Linux poll 함수 사용하여 IPC data 수신될 때까지 지속대기, 실시간 송수신을 위해서 다른 것 사용.
 
         if (poll_ret == -1) {
             perror("Poll failed");
@@ -469,7 +487,7 @@
                     cmd1 = ((uint16_t)readBuf[3]<<8)|((uint16_t)readBuf[4]);
                     cmd2 = ((uint16_t)readBuf[5]<<8)|((uint16_t)readBuf[6]);
                     IPC_datalen = ((uint16_t)readBuf[7]<<8)|((uint16_t)readBuf[8]);
-                    printf("cmd1: 0x%x,cmd2: 0x%x, length: 0x%x\n",cmd1,cmd2,IPC_datalen);
+                    printf("cmd1: 0x%x, cmd2: 0x%x, length: 0x%x\n",cmd1,cmd2,IPC_datalen);
 					// CM7-Core에서 보낸 데이터가 동일하면 read하여 CAN data 파싱
                     if((cmd1 == TCC_IPC_CMD_AP_TEST)&&(cmd2 == TCC_IPC_CMD_AP_SEND))
                     {
@@ -689,12 +707,24 @@
 }
 // IPC Header + Data send to Driver
 int32_t ipc_write3(char *wbuffer, size_t wSize)
-{
+{	
+	/* IPC CAN(FD) Format
+		0xFF 0x55 0xAA			: IPC Header
+		0x0 0xFF				: IPC cmd1
+		0x0 0x1					: IPC cmd2(port number)
+		0x0 0x15				: IPC data size(can header + payload)
+		0x0 0x50 0x0 0x0 0x10	: CAN Header(Timestamp on, Protocol, CAN ID, FDF, RTR, BRS)
+		0x50 0x51 0x52 0x53		: Payload
+		0xCF 0xCD				: CRC
+	 */
 	int32_t ret =-1;
 
 	if((fd3 > -1)&&(wbuffer != NULL)&&(wSize > (size_t)0))
 	{
-		ret = write(fd3, wbuffer, wSize); // Trasmit IPC data to the driver as much as the size passed as a Parameter 
+		printf("ipc_write3 start\n"); // 241007 sy.kim
+
+		// Trasmit IPC data to the driver as much as the size passed as a Parameter 
+		ret = write(fd3, wbuffer, wSize); 
 		/*for(i=0; i<wSize; i++) {
 			printf("writeBuf[%d]: 0x%lx\n", i, wbuffer[i]);
 		}*/
@@ -1027,7 +1057,8 @@
 
 	if(wbuffer != NULL)
 	{
-		(void)printf("[INFO] A65E send CAN to M7_NP\n");
+		// 241003 sy.kim
+		(void)printf("[INFO] ipc_make_packet: A65E send CAN to M7_NP\n");
 		// IPC Header
 		wbuffer[0] = 0xFF;
 		wbuffer[1] = 0x55;
@@ -1055,6 +1086,7 @@
 	return packet_size;
 }
 
+
 int ipc_Lpa_packet(char *wbuffer, char *ipc_buff, int ipcCmd1, int ipcCmd2, int data_length)
 {
 	int32_t i = 0;
@@ -1063,6 +1095,9 @@
 
 	if(wbuffer != NULL)
 	{
+		// 241007 sy.kim
+		(void)printf("[INFO] ipc_Lpa_packet: A65E send CAN to M7_NP\n");
+
 		wbuffer[0] = 0xFF;
 		wbuffer[1] = 0x55;
 		wbuffer[2] = 0xAA;
@@ -1077,11 +1112,19 @@
 		for(i = 9; i < packet_size - 2; i++) {
 			wbuffer[i] = ipc_buff[i-9];
 		}
-
+		
 		crc = IPC_CalcCrc16(&wbuffer[0], (packet_size - 2U), 0);
 
 		wbuffer[i++] = (char) ((crc & (int) 0xFF00) >> (int) 8);
 		wbuffer[i] = (char) (crc & (int) 0x00FF);
+
+		// 241007 sy.kim
+		printf("IPC LPA packet Data ");
+		for(int j=0; j<packet_size; j++)
+		{
+			printf("0x%X ", wbuffer[j]);
+		}
+		printf("\n");
 	}
 
 	return packet_size;
@@ -1105,9 +1148,10 @@
     }
     else // Standard ID
     {
-		(void)printf("build_CANHeader() : Standard ID, ");
+		(void)printf("build_CANHeader() : Standard ID, "); // 241003 sy.kim
         if(uFDF == 1U) // CAN FD
         {
+			(void)printf("CAN FD\n"); // 241003 sy.kim
             can_header_frame = TIMESTAMP(Timestamp_onoff) + PROTOCOL(0U) + CANID(uCAN_ID) + FDF(1U) + RTR(0U) + IDE(0U) + BRS(uBRS);
         }
         else // classic CAN
@@ -1121,17 +1165,17 @@
     header_buffer[2] = (uint8_t)((can_header_frame&0x0000FF0000ULL)>>16U);
     header_buffer[3] = (uint8_t)((can_header_frame&0x00FF000000ULL)>>24U);
     header_buffer[4] = (uint8_t)((can_header_frame&0xFF00000000ULL)>>32U);
-
 }
 
 
 void LPA_msg(uint8_t protocol, uint8_t port, uint32_t id, uint8_t ext_opMode, uint8_t data_len, uint8_t *pData)
 {
 
-	if ((pData != NULL) && ( data_len <= 8U))
+	if ((pData != NULL) && (data_len <= 8U))
 	{
         if(protocol == PROTOCOL_CAN)
         {
+			(void)printf("CAN LPA_msg()\n"); // 241003 sy.kim
 			build_CANHeader(&CetracSend[0], TIMESTAMP_ON, (uint32_t)id, 0U, ext_opMode, 0U);
         }
         else
@@ -1139,7 +1183,29 @@
         	/**/
         }
 		memcpy(&CetracSend[LPA_TX_HDR_SIZE], pData, data_len);
+	}
+	// 241007 sy.kim
+	else if ((pData != NULL) && (data_len <= 64U) )
+	{
+        if(protocol == PROTOCOL_CAN)
+        {
+			// 241003 sy.kim
+			(void)printf("CAN FD LPA_msg()\n");
+			build_CANHeader(&CetracSend[0], TIMESTAMP_ON, (uint32_t)id, 1U, ext_opMode, 0U);
+        }
+        else
+        {
+        	/**/
+        }
+		memcpy(&CetracSend[LPA_TX_HDR_SIZE], pData, data_len); // Add payload after Header 
 
+		// 241007 sy.kim
+		printf("CetracSend Data Header ");
+		for(int i=0; i<5; i++)
+		{
+			printf("0x%X ", CetracSend[i]);
+		}
+		printf("\n");
 	}
 }
 
@@ -1247,12 +1313,32 @@
 	        (void)ipc_write2(&writeBuf, ipc_packet_size);
 		}
     }
-	else if(!strncmp(cmd, "wr3", 3)) {
-		for (i = 0; i < send_num; i++)
+	else if(!strncmp(cmd, "wr3", 3)) { 
+		// 241007 sy.kim
+		// for (i = 0; i < send_num; i++)
+		// {
+		// 	ipc_packet_size = ipc_make_packet(writeBuf, i, ipc_cmd1, ipc_cmd2, ipc_data_length);
+	    //     (void)ipc_write3(writeBuf, ipc_packet_size);
+		// }
+		IPCLen = RX_data_len + LPA_TX_HDR_SIZE;
+		
+		printf("Write CAN(FD) ID 0x%X, Port %d, Len %d\n", RX_can_ID, RX_can_port, RX_data_len);
+		printf("Write CAN(FD) Data: ");
+		for(i = 0U; i < RX_data_len; i++)
 		{
-			ipc_packet_size = ipc_make_packet(writeBuf, i, ipc_cmd1, ipc_cmd2, ipc_data_length);
-	        (void)ipc_write3(writeBuf, ipc_packet_size);
+			printf("%02x ", RX_Data[i]);
 		}
+		printf("\n");
+
+		(void)LPA_msg(PROTOCOL_CAN, RX_can_port, RX_can_ID, 0, (uint8_t)RX_data_len, &RX_Data);
+		
+		ipc_packet_size = ipc_Lpa_packet(writeBuf, CetracSend, TCC_IPC_CMD_AP_TEST, RX_can_port, IPCLen);
+		// 241104 sy.kim IPC performance test 
+		gettimeofday(&start, NULL);
+		(void)ipc_write3(writeBuf, ipc_packet_size);
+		gettimeofday(&end, NULL);
+		elapsed_time = (end.tv_sec - start.tv_sec) * 1000000L + (end.tv_usec - start.tv_usec);
+		printf("Execution Time: %ld us\n", elapsed_time);
     }
     else if(!strncmp(cmd, "can", 3)) {
 
@@ -1284,7 +1370,47 @@
 		ipc_packet_size = ipc_Lpa_packet(writeBuf, CetracSend, TCC_IPC_CMD_AP_TEST, portN, IPCLen);
 		(void)ipc_write3(writeBuf, ipc_packet_size);
 
-    }	
+    }
+	// 241003 sy.kim
+	else if(!strncmp(cmd, "canfd", 3)) {
+
+		DataLen = RX_data_len; 
+		IPCLen = DataLen+LPA_TX_HDR_SIZE;
+		for(i=0; i<DataLen; i++)
+		{
+			Data[i]= i;
+		}
+
+		memset(str, 0, sizeof(str));
+
+		printf("CANFD channel : 1 ~ 16 \n");
+		fgets(str , 128, stdin);
+		(void)sscanf(str,"%d",&portN);
+
+		printf("CANFD ID : 1 ~ 7ff \n");
+		fgets(str , 128, stdin);
+		(void)sscanf(str,"%x",&CanID);
+
+		printf("port Number :%d, CANFD ID : 0x%x, Data length : %d\n", portN, CanID, DataLen);
+		(void)LPA_msg(PROTOCOL_CAN, portN, CanID, 0, (uint8_t)DataLen, Data);
+
+		for(i=0; i<IPCLen; i++)
+		{
+			printf("senddata[%d]:%d\n", i, CetracSend[i]);
+		}
+
+		/* Driver Open 이후 ipc_make_packet과 ipc_srite.를 같이 호출하여 parameter만 전달하면 CM7-Core로 IPC Data 전달 가능함
+			writeBuf			: IPC total data buffer
+			CetracSend			: CM7-core에 전달할 IPC Data buffer 에제코드 파라미터 변경함
+			TCC_IPC_CMD_AP_TEST	: IPC cmd1 수신할 CM7-Core와 동일한 값이어야함
+			portN				: IPC cmd2
+			IPCLen				: CM7-Core에 전달할 IPC Data size 
+		 */
+		ipc_packet_size = ipc_Lpa_packet(writeBuf, CetracSend, TCC_IPC_CMD_AP_TEST, portN, IPCLen);
+		(void)ipc_write3(writeBuf, ipc_packet_size);
+
+    }
+	
 	else if(!strncmp(cmd, "rd", 3)) {
 		read_size0 = read(fd0, &readBuf0[0], ipc_read_packet_size);
 		read_size1 = read(fd1, &readBuf1[0], ipc_read_packet_size);
@@ -1454,6 +1580,8 @@
 		(void)printf("set write : Enter wr2 --> CM7-2\n");
 		(void)printf("set write : Enter wr3 --> CM7-np\n");
 		(void)printf("can write : Enter can --> CM7-np\n");
+		// 241003 sy.kim
+		(void)printf("can write : Enter canfd --> CM7-np\n");
 
         (void)printf("set read : Enter rd --> CM7-ALL\n");
 		(void)printf("set read : Enter rd0 --> CM7-0\n");
